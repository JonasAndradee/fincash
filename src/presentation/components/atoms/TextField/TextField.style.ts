import { ReactNode } from 'react'
import styled, { css } from 'styled-components'
import theme from '../../../theme'

const Container = styled.div``

const LabelContent = styled.div`
  display: flex;
  align-items: center;
  margin-bottom: ${theme.space('micro')};
`

const Label = styled.span`
  ${theme.typography.small}
  color: ${theme.color('coal')};
`

interface IInputContent {
  selected?: boolean
  error?: ReactNode | string
  disabled?: boolean
}

const InputContent = styled.div<IInputContent>`
  display: flex;
  align-items: center;
  justify-content: center;

  padding: 0.75rem 1rem;

  border: 2px solid;
  border-color: ${({ error, selected }) => {
    if (selected) return theme.color('steel', 'dark')
    if (error) return theme.color('error')
    return theme.color('steel')
  }};
  border-radius: 8px;
  
  background-color: ${theme.color('white')};
  ${({ disabled }) => disabled && css`
    background-color: ${theme.color('silver')};
  `}
`

const Input = styled.input`
  width: 100%;
  border: none;

  ${theme.typography.normal}
  color: ${theme.color('coal', 'dark')};
  background: transparent;

  ::placeholder {
    color: ${theme.color('steel', 'dark')};
  }

  :invalid {
    box-shadow: none;
  }

  :focus {
    outline: none;
  }

  ${({ disabled }) => disabled && css`
    background-color: transparent;
    cursor: not-allowed;
    pointer-events: none;

    ::placeholder {
      color: ${theme.color('steel', 'dark')};
    }
  `}
`

const MessageContent = styled.div`
  margin-top: ${theme.space('micro')};
  display: inline-flex;

  svg{
    margin-right: ${theme.space('micro')};
  }
`

const MessageLabel = styled.span`
  ${theme.typography.small}
  color: ${theme.color('steel', 'dark')};
`

export default {
  Container,
  LabelContent,
  Label,
  InputContent,
  Input,
  MessageContent,
  MessageLabel
}
