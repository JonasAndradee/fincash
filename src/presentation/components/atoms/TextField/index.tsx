import React, { memo, ReactNode } from 'react'

import { MdErrorOutline, MdWarning } from 'react-icons/md'
import { CgAsterisk } from 'react-icons/cg'

import Style from './TextField.style'
import theme from '../../../theme'

interface IProps {
  id?: string,
  label?: string,
  placeholder?: string,
  value?: string,
  icon?: ReactNode,
  required?: boolean,
  selected?: boolean,
  disabled?: boolean,
  error?: ReactNode | string,
  alert?: ReactNode,
  type?: string
}

const TextField: React.FC<IProps> = props => {
  const {
    id, required, selected, disabled,
    icon, label, value, placeholder,
    error, alert, ...other
  } = props

  return (
    <Style.Container id={id}>
      {
        label && (
          <Style.LabelContent>
            <Style.Label>{label}</Style.Label>
            {required && (<CgAsterisk color={theme.color('red')}size={16} />)}
          </Style.LabelContent>
        )
      }
      <Style.InputContent selected={selected} error={error} disabled={disabled}>
        {icon}
        <Style.Input value={value} placeholder={placeholder} disabled={disabled} {...other} />
      </Style.InputContent>
      {
        error && (
          <Style.MessageContent>
            <MdErrorOutline size={16} color={theme.color('error')} />
            <Style.MessageLabel>{error}</Style.MessageLabel>
          </Style.MessageContent>
        )
      }
      {
        alert && (
          <Style.MessageContent>
            <MdWarning size={19} color={theme.color('warning')} />
            <Style.MessageLabel>{alert}</Style.MessageLabel>
          </Style.MessageContent>
        )
      }
    </Style.Container>
  )
}

export default memo(TextField)
