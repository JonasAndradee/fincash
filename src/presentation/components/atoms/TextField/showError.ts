type Meta = {
  touched?: boolean
  error?: string
  submitError?: string
  dirtySinceLastSubmit?: boolean
}

function showError (meta: Meta) {
  const {
    touched, error, submitError, dirtySinceLastSubmit
  } = meta
  if (!touched || (dirtySinceLastSubmit)) return ''

  return (error || submitError || '')
}

export default showError
