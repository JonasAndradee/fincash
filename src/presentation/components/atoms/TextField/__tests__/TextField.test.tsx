import React from 'react'
import renderer from 'react-test-renderer'
import TextField from '..'
import showError from '../showError'

const TextFieldWithTheme = TextField
const label = 'Email'
const placeholder = 'peter@boticario.com'
const types = ['email', 'password', 'text']
const icon = <svg></svg>
const value = 'venom@boticario.com'

const errorString = 'O nome deve ter mais de 3 caracteres.'
const ErrorNode = () => <div>Valor inválido. Consulte <a href="/">aqui</a> para mais informações</div>

const alertString = 'Nome é a forma como chamamos alguém.'
const AlertNode = () => <div>Você quis dizer <a href="/">peter@boticario.com</a>?</div>

/** showError function */
const metaWithoutTouched = { touched: false }
const metaWithTouched = { touched: true }
const metaWithLastEdited = { dirtySinceLastSubmit: true }
const metaWithError = { touched: true, error: 'the third movie of Sam Raimi is a mess' }
const metaWithSubmitError = { touched: true, submitError: 'very bad' }

describe('given: TextField', () => {
  describe('checking: properties', () => {
    test('render a TextField with label when label is not empty', () => {
      const snap = renderer.create(<TextFieldWithTheme label={label} />).toJSON()
      expect(snap).toMatchSnapshot()
    })
    test('render a TextField with placeholder when placeholder is not empty', () => {
      const snap = renderer.create(<TextFieldWithTheme placeholder={placeholder} />).toJSON()
      expect(snap).toMatchSnapshot()
    })
    test('render a TextField with icon before input when icon is not empty', () => {
      const snap = renderer.create(<TextFieldWithTheme icon={icon} />).toJSON()
      expect(snap).toMatchSnapshot()
    })
  })
  describe('checking: state properties', () => {
    test('render a simple TextField when do not receive props', () => {
      const snap = renderer.create(<TextFieldWithTheme />).toJSON()
      expect(snap).toMatchSnapshot()
    })
    test('render a TextField with label and required when label is not empty and required is true', () => {
      const snap = renderer.create(<TextFieldWithTheme label={label} required />).toJSON()
      expect(snap).toMatchSnapshot()
    })
    test('render a TextField with value when value is not empty', () => {
      const snap = renderer.create(<TextFieldWithTheme value={value} />).toJSON()
      expect(snap).toMatchSnapshot()
    })
    test('render a TextField with type when type is not empty', () => {
      const snapEmail = renderer.create(<TextFieldWithTheme type={types[0]} />).toJSON()
      const snapPass = renderer.create(<TextFieldWithTheme type={types[1]} />).toJSON()
      const snapText = renderer.create(<TextFieldWithTheme type={types[2]} />).toJSON()
      expect(snapEmail).toMatchSnapshot()
      expect(snapPass).toMatchSnapshot()
      expect(snapText).toMatchSnapshot()
    })
    test('render a selected TextField when selected is true', () => {
      const snap = renderer.create(<TextFieldWithTheme value={value} selected />).toJSON()
      expect(snap).toMatchSnapshot()
    })
    test('render a disabled TextField when disabled is true', () => {
      const snap = renderer.create(<TextFieldWithTheme value={value} disabled />).toJSON()
      expect(snap).toMatchSnapshot()
    })
  })
  describe('checking: hint (bottom text) properties', () => {
    test('render a TextField with error hint when error is not null', () => {
      const snapWithString = renderer.create(
        <TextFieldWithTheme error={errorString} />
      ).toJSON()
      const snapWithNode = renderer.create(
        <TextFieldWithTheme error={<ErrorNode />} />
      ).toJSON()
      expect(snapWithString).toMatchSnapshot()
      expect(snapWithNode).toMatchSnapshot()
    })
    test('render a TextField with alert hint when alert is not null', () => {
      const snapWithString = renderer.create(
        <TextFieldWithTheme alert={alertString} />
      ).toJSON()
      const snapWithNode = renderer.create(
        <TextFieldWithTheme alert={<AlertNode />} />
      ).toJSON()
      expect(snapWithString).toMatchSnapshot()
      expect(snapWithNode).toMatchSnapshot()
    })
  })
  describe('checking: showError function', () => {
    test('returns empty when is not touched yet or is modified before re-submit or is touched but dont have error', () => {
      expect(showError(metaWithoutTouched)).toEqual('')
      expect(showError(metaWithLastEdited)).toEqual('')
      expect(showError(metaWithTouched)).toEqual('')
    })
    test('returns the message error when has any error in meta and should render error', () => {
      expect(showError(metaWithError)).toEqual(metaWithError.error)
      expect(showError(metaWithSubmitError)).toEqual(metaWithSubmitError.submitError)
    })
  })
})
