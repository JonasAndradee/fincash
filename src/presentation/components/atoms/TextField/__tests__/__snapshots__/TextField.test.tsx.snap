// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given: TextField checking: hint (bottom text) properties render a TextField with alert hint when alert is not null 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
    />
  </div>
  <div
    className="sc-jSgvzq uwifr"
  >
    <svg
      color="#7C3AED"
      fill="currentColor"
      height={19}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": "#7C3AED",
        }
      }
      viewBox="0 0 24 24"
      width={19}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
      />
    </svg>
    <span
      className="sc-gKseQn fbmicb"
    >
      Nome é a forma como chamamos alguém.
    </span>
  </div>
</div>
`;

exports[`given: TextField checking: hint (bottom text) properties render a TextField with alert hint when alert is not null 2`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
    />
  </div>
  <div
    className="sc-jSgvzq uwifr"
  >
    <svg
      color="#7C3AED"
      fill="currentColor"
      height={19}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": "#7C3AED",
        }
      }
      viewBox="0 0 24 24"
      width={19}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
      />
    </svg>
    <span
      className="sc-gKseQn fbmicb"
    >
      <div>
        Você quis dizer 
        <a
          href="/"
        >
          peter@boticario.com
        </a>
        ?
      </div>
    </span>
  </div>
</div>
`;

exports[`given: TextField checking: hint (bottom text) properties render a TextField with error hint when error is not null 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU jNqtMA"
  >
    <input
      className="sc-eCstlR cVJSuX"
    />
  </div>
  <div
    className="sc-jSgvzq uwifr"
  >
    <svg
      color="#EF4444"
      fill="currentColor"
      height={16}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": "#EF4444",
        }
      }
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
      />
    </svg>
    <span
      className="sc-gKseQn fbmicb"
    >
      O nome deve ter mais de 3 caracteres.
    </span>
  </div>
</div>
`;

exports[`given: TextField checking: hint (bottom text) properties render a TextField with error hint when error is not null 2`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU jNqtMA"
  >
    <input
      className="sc-eCstlR cVJSuX"
    />
  </div>
  <div
    className="sc-jSgvzq uwifr"
  >
    <svg
      color="#EF4444"
      fill="currentColor"
      height={16}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": "#EF4444",
        }
      }
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
      />
    </svg>
    <span
      className="sc-gKseQn fbmicb"
    >
      <div>
        Valor inválido. Consulte 
        <a
          href="/"
        >
          aqui
        </a>
         para mais informações
      </div>
    </span>
  </div>
</div>
`;

exports[`given: TextField checking: properties render a TextField with icon before input when icon is not empty 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <svg />
    <input
      className="sc-eCstlR cVJSuX"
    />
  </div>
</div>
`;

exports[`given: TextField checking: properties render a TextField with label when label is not empty 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-gsTEea lcSwzB"
  >
    <span
      className="sc-dlfnuX evOzTz"
    >
      Email
    </span>
  </div>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
    />
  </div>
</div>
`;

exports[`given: TextField checking: properties render a TextField with placeholder when placeholder is not empty 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
      placeholder="peter@boticario.com"
    />
  </div>
</div>
`;

exports[`given: TextField checking: state properties render a TextField with label and required when label is not empty and required is true 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-gsTEea lcSwzB"
  >
    <span
      className="sc-dlfnuX evOzTz"
    >
      Email
    </span>
    <svg
      color="#FF3859"
      fill="none"
      height={16}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": "#FF3859",
        }
      }
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
        fill="currentColor"
      />
    </svg>
  </div>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
    />
  </div>
</div>
`;

exports[`given: TextField checking: state properties render a TextField with type when type is not empty 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
      type="email"
    />
  </div>
</div>
`;

exports[`given: TextField checking: state properties render a TextField with type when type is not empty 2`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
      type="password"
    />
  </div>
</div>
`;

exports[`given: TextField checking: state properties render a TextField with type when type is not empty 3`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
      type="text"
    />
  </div>
</div>
`;

exports[`given: TextField checking: state properties render a TextField with value when value is not empty 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
      value="venom@boticario.com"
    />
  </div>
</div>
`;

exports[`given: TextField checking: state properties render a disabled TextField when disabled is true 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU caAZRo"
    disabled={true}
  >
    <input
      className="sc-eCstlR cqCdiK"
      disabled={true}
      value="venom@boticario.com"
    />
  </div>
</div>
`;

exports[`given: TextField checking: state properties render a selected TextField when selected is true 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU lhqIjs"
    selected={true}
  >
    <input
      className="sc-eCstlR cVJSuX"
      value="venom@boticario.com"
    />
  </div>
</div>
`;

exports[`given: TextField checking: state properties render a simple TextField when do not receive props 1`] = `
<div
  className="sc-bdfBQB hsUKZc"
>
  <div
    className="sc-hKgJUU gFsRuv"
  >
    <input
      className="sc-eCstlR cVJSuX"
    />
  </div>
</div>
`;
